### PR TITLE
Feature/set up associations

### PR DESCRIPTION
**Pull request by Shedrack-Sunday**
## Changes Made:

-- Created a class Classroom with the following:
@label instance variable initialized in the constructor.
Setter and getter for @label (remember about attr_accessor).

-- Created the has-many/belongs-to relationship between Classroom and Student. The following is  implemented:
The has-many side (a classroom has many students).
A  belongs-to side (a student belongs to a classroom).

-- Created a class Book with the following:
@title and @author instance variables initialized in the constructor.
Setters and getters for instance variables (remember about attr_accessor).

-- Created a class Rental with the following:
@date instance variable  initialized in the constructor.
Setter and getter for @date (remember about attr_accessor).A many-to-many (also has-many-through) relationship between Person and Book using the intermediate class Rental. The following is  implemented:
- A has-many side of Book and Rental (a book has many rentals).
- The belongs-to side of Rental and Book (a rental belongs to a book).
- The has-many side of Person and Rental (a person has many rentals).
- The belongs-to side of Rental and Person (a rental belongs to a person).

### CODE REVIEW

File: classroom.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+class Classroom
+  attr_reader :label, :students
```
Comment (OLIPLICHE) [inline classroom.rb]: - [ ] Reading from your project requirements: **Create a class Classroom with the following**
- 👉  "Setter and getter for @label (remember about attr_accessor)".
[Please Open this Link to learn more](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_principles_examples.md#useful-shortcuts-attr_reader-attr_writer-and-attr_accessor)

File: student.rb [hunk 2]
```diff
@@ -1,9 +1,11 @@
 require './person'
+require './classroom'
 
 class Student < Person
   def initialize(age, classroom, name: 'Unknown', parent_permission: true)
     super(age, name, parent_permission: parent_permission)
     @classroom = classroom
+    classroom.add_student(self)
```
Comment (OLIPLICHE) [inline student.rb:10]: - [ ] From the requirements: "Create the has-many/belongs-to relationship between Classroom and Student":
- 👉  Make sure that when setting the classroom for a student it also adds it to the classrooms' students.
- [Please open this link and Read the example provided.](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md#both-ways-relationships)

File: book.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+class Book
+  attr_reader :rentals
+  attr_accessor :title, :author
```
Comment (OLIPLICHE) [inline book.rb]: - rentals have both Getters and setters methods (`attr_reader` and `attr_writer` respectively ). Therefore remove it to where it is and move it to `attr_accessor`. 
-[ Read More here](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_principles_examples.md#useful-shortcuts-attr_reader-attr_writer-and-attr_accessor)

File: book.rb [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+
+  def initialize(title, author)
+    @title = title
+    @author = author
+    @rentals = []
+  end
```
Comment (OLIPLICHE) [inline book.rb:8]: - [ ] **A book has many rentals**: A method for adding rentals should be present. 
- Please [Open this link and see examples](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md#many-to-many-relationship)
- something like this:
```
def add_rental(person, date)
    Rental.new(date, self, person)
  end
```

File: rental.rb [hunk 5]
```diff
@@ -0,0 +1,11 @@
+class Rental
+  attr_accessor :date, :book, :person
+
+  def initialize(date)
```
Comment (OLIPLICHE) [inline rental.rb]: - Initialize `book` and `person` as well. the same way you did with date

File: person.rb [hunk 6]
```diff
@@ -1,5 +1,11 @@
...
-  attr_reader :id, 
+require './nameable'
+require './capitalize_decorator'
+require './trimmer_decorator'
+require './book'
+require './rental'
+
+class Person < Nameable
+  attr_reader :id, :rentals
   attr_accessor :name, :age
```
Comment (OLIPLICHE) [inline person.rb]: - Move rentals to `attr_accessor`

File: person.rb [hunk 7]
```diff
@@ -8,16 +14,18 @@ def initialize(age, name = 'Unknown', parent_permission: true)
...
 
   def of_age?
     @age >= 18
   end
 
-  public
-
   def can_use_services?
     of_age? || @parent_permission
   end
```
Comment (OLIPLICHE) [inline person.rb:32]: - [ ] **A person has many rentals**: A method for adding rentals should be present
- Refer to my previous comment in book.rb

**Review by OLIPLICHE — CHANGES_REQUESTED** [review]
Hi @Shedrack-Sunday, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
- Linter is green (the last commit has a green check mark (✅)) 👏
- README file is descriptive 👏
- Commit messages are meaningful 👏 👏
- Gitflow is correctly used (Pull Request is made from the feature branch into the dev branch) 👏
## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by KayLemba — APPROVED** [review]
[# STATUS: APPROVED :green_circle: :clap: 

Hello  @Shedrack-Sunday     , 
Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

HIGHLIGHTS:

- [x]  Your repo is professional 
- [x] Descriptive P.R





 








Now you can merge this branch and move on :1st_place_medal: 
Happy Coding cheers 🥇 💯 



well-1

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._]

**Pull request merged**